extensions/Worker.Extensions.EventGrid/src/TypeConverters/EventGridBinaryDataConverter.cs

```diff
@@ -0,0 +1,97 @@
+    [SupportedConverterType(typeof(BinaryData[]))]
+    internal class EventGridBinaryDataConverter : IInputConverter
+    {
+        private readonly IOptions<WorkerOptions> _workerOptions;
+
+        public EventGridBinaryDataConverter(IOptions<WorkerOptions> workerOptions)
+        {
+            _workerOptions = workerOptions ?? throw new ArgumentNullException(nameof(workerOptions));
+        }
+        public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.TargetType != typeof(BinaryData) && context.TargetType != typeof(BinaryData[]))
+            {
+                return new(ConversionResult.Unhandled());
+            }
+
+            try
+            {
+                var contextSource = context?.Source as string;
+
+                if (contextSource is null)
+                {
+                    return new(ConversionResult.Unhandled());
+                }
+
+                if (context!.TargetType == typeof(BinaryData))
+                {
+                    var binaryData = BinaryData.FromObjectAsJson<string>(contextSource);
+                    return new(ConversionResult.Success(binaryData));
+                }
+                else if (context!.TargetType == typeof(BinaryData[]))
+                {
+                    byte[] byteArray = Encoding.UTF8.GetBytes(contextSource);
+                    MemoryStream stream = new MemoryStream(byteArray);
+
+                    var jsonData = (List<object>?)_workerOptions?.Value?.Serializer?.Deserialize(stream, typeof(List<object>), CancellationToken.None);
+                    List<BinaryData> binaryDataList = new List<BinaryData>();
+
+                    if (jsonData is not null)
+                    {
+                        foreach (var item in jsonData)
+                        {
+                            if (item is not null)
+                            {
+                                var binaryData = new BinaryData(item);
+                                binaryDataList.Add(binaryData);
+                            }
+                        }
+                        return new(ConversionResult.Success(binaryDataList.ToArray()));
+                    }
+                }
+            }
+            catch (JsonException ex)
+            {
+                string msg = String.Format(CultureInfo.CurrentCulture,
+                    @"Binding parameters to complex objects uses Json.NET serialization.
+                    1. Bind the parameter type as 'string' instead to get the raw values and avoid JSON deserialization, or
+                    2. Change the queue payload to be valid json.
+                    The JSON parser failed: {0}",
+                    ex.Message);
+
+                return new(ConversionResult.Failed(new InvalidOperationException(msg)));
```

Comment:
Should we include the underlying JSON exception?

Reply:
yup it's included already by adding in ex.message!

Reply:
It's included in the message, but should we also include it as an inner exception for the InvalidOperationException?

Reply:
Flagging this as unresolved as it doesn't look like the change was applied.

The exception should indeed be added as the inner exception for the `InvalidOperationException` being created. We can also remove the message from the one you're crafting